File: HydroCalc/HydroCalculator.swift
import Foundation

struct HydroCalculator {
    static let waterFactor = 1.163 // Wh/(l·K), practical approximation for heating water

    static func flowLitersPerHour(powerKW: Double, deltaT: Double) -> Double? {
        guard powerKW > 0, deltaT > 0 else { return nil }
        return powerKW * 1000 / (waterFactor * deltaT)
    }

    static func powerKW(flowLh: Double, deltaT: Double) -> Double? {
        guard flowLh > 0, deltaT > 0 else { return nil }
        return flowLh * waterFactor * deltaT / 1000
    }

    static func deltaT(powerKW: Double, flowLh: Double) -> Double? {
        guard powerKW > 0, flowLh > 0 else { return nil }
        return powerKW * 1000 / (flowLh * waterFactor)
    }

    static func velocity(flowLh: Double, innerDiameterMM: Double) -> Double? {
        guard flowLh > 0, innerDiameterMM > 0 else { return nil }
        let flowM3s = flowLh / 3_600_000
        let diameterM = innerDiameterMM / 1000
        let area = Double.pi * pow(diameterM, 2) / 4
        return flowM3s / area
    }

    static func innerDiameterMM(flowLh: Double, targetVelocity: Double) -> Double? {
        guard flowLh > 0, targetVelocity > 0 else { return nil }
        let flowM3s = flowLh / 3_600_000
        let diameterM = sqrt(4 * flowM3s / (Double.pi * targetVelocity))
        return diameterM * 1000
    }

    static func kv(flowM3h: Double, differentialPressureBar: Double) -> Double? {
        guard flowM3h > 0, differentialPressureBar > 0 else { return nil }
        return flowM3h / sqrt(differentialPressureBar)
    }

    static func differentialPressureBar(flowM3h: Double, kv: Double) -> Double? {
        guard flowM3h > 0, kv > 0 else { return nil }
        return pow(flowM3h / kv, 2)
    }
}

extension Double {
    var hydroInputString: String {
        formatted(.number.precision(.fractionLength(0...3)))
    }
}

extension String {
    var hydroDouble: Double? {
        Double(replacingOccurrences(of: ",", with: "."))
    }
}

File: android/app/src/main/java/de/kamilunavo/hydrocalc/HydroCalculator.kt
package de.kamilunavo.hydrocalc

import kotlin.math.PI
import kotlin.math.pow
import kotlin.math.sqrt

object HydroCalculator {
    const val WATER_FACTOR = 1.163

    fun flowLitersPerHour(powerKW: Double, deltaT: Double): Double? {
        if (powerKW <= 0 || deltaT <= 0) return null
        return powerKW * 1000.0 / (WATER_FACTOR * deltaT)
    }

    fun powerKW(flowLh: Double, deltaT: Double): Double? {
        if (flowLh <= 0 || deltaT <= 0) return null
        return flowLh * WATER_FACTOR * deltaT / 1000.0
    }

    fun deltaT(powerKW: Double, flowLh: Double): Double? {
        if (powerKW <= 0 || flowLh <= 0) return null
        return powerKW * 1000.0 / (flowLh * WATER_FACTOR)
    }

    fun velocity(flowLh: Double, innerDiameterMM: Double): Double? {
        if (flowLh <= 0 || innerDiameterMM <= 0) return null
        val flowM3s = flowLh / 3_600_000.0
        val diameterM = innerDiameterMM / 1000.0
        val area = PI * diameterM.pow(2) / 4.0
        return flowM3s / area
    }

    fun innerDiameterMM(flowLh: Double, targetVelocity: Double): Double? {
        if (flowLh <= 0 || targetVelocity <= 0) return null
        val flowM3s = flowLh / 3_600_000.0
        val diameterM = sqrt(4.0 * flowM3s / (PI * targetVelocity))
        return diameterM * 1000.0
    }

    fun kv(flowM3h: Double, differentialPressureBar: Double): Double? {
        if (flowM3h <= 0 || differentialPressureBar <= 0) return null
        return flowM3h / sqrt(differentialPressureBar)
    }

    fun differentialPressureBar(flowM3h: Double, kv: Double): Double? {
        if (flowM3h <= 0 || kv <= 0) return null
        return (flowM3h / kv).pow(2)
    }
}
